### PR TITLE
Add backwards compatibility for Jacoco 0.7.4 and below

### PR DESCRIPTION
**Pull request by MarkRx**
Jacoco version 0.7.5 made a breaking change to the internal exec file. This propagated to the Jenkins jacoco plugin causing it to not work with jacoco exec files that were created from jacoco 0.7.4 or below (#55). This breaking change presents a major hurdle for large Jenkins instances as it requires a mass coordinated upgrade of jacoco versions in order to upgrade the Jenkins plugin (and vice-versa).

This change adds backwards compatibility for Jacoco v0.7.4 and below (exec version 0x1006) using a technique suggested by @cardil at https://github.com/jenkinsci/jacoco-plugin/pull/55#issuecomment-158204390. It uses delegates for Analyzer and ExecutationDataReader that call the current version of jacoco or a shadowed copy based on the version of the exec file. It will allow the Jenkins plugin to be able to consume both old and new jacoco files.

### CODE REVIEW

File: src/main/java/org/jacoco/core/JacocoUtil.java [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+
+public class JacocoUtil {
+	public static final char VERSION_074 = 0x1006;
+	
+	public static char getVersion(File jacocoExec) throws IOException {
+		char version;
+		
+		try (DataInputStream in = new DataInputStream(new FileInputStream(jacocoExec))) {
+			if (in.read() != ExecutionDataWriter.BLOCK_HEADER) {
+				throw new IOException("Invalid execution data file.");
```
Comment (mheinzerling) [inline src/main/java/org/jacoco/core/JacocoUtil.java]: Can we give more details for debugging here?

File: src/main/java/org/jacoco/core/JacocoUtil.java [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+			}
+			
+			if (in.readChar() != ExecutionDataWriter.MAGIC_NUMBER) {
+				throw new IOException("Invalid execution data file.");
```
Comment (mheinzerling) [inline src/main/java/org/jacoco/core/JacocoUtil.java]: Can we give more details for debugging here?

File: src/test/java/hudson/plugins/jacoco/ExecutionFileLoaderTest.java [hunk 3]
```diff
@@ -115,6 +115,26 @@ public void testLoadBundleWithExecFile() throws IOException {
...
     
     @Test
+    public void testLoadBundleWithExecFileVersion074() throws IOException {
+        ExecutionFileLoader loader = new ExecutionFileLoader();
+        
+        assertTrue("This test requires that a jacoco.exec file exists in the target-directory", 
+                new File("target/jacoco.exec").exists());
+        loader.setClassDir(new FilePath(new File("target/classes")));
+        loader.setExcludes("excludme.test");
+        loader.addExecFile(new FilePath(new File("src/test/jacoco_074.exec")));
```
Comment (mheinzerling) [inline src/test/java/hudson/plugins/jacoco/ExecutionFileLoaderTest.java]: Didn't checked it out yet, but is there a "resources" missing here?

**Comment by MarkRx** [comment]
Made CR changes.

**Comment by MarkRx** [comment]
Any updates on this?

**Comment by saimad** [comment]
Yes, we also need this. thanks MarkRx for attempting to fix this. It will be great to merge this in and release a new plugin version... Pretty please!

**Comment by saimad** [comment]
@bootstraponline ^^^ please see if this can be merged.

**Comment by bootstraponline** [comment]
??

I'm not a dev on this project. I've moved to using bitrise and buddybuild instead of jenkins.

**Comment by saimad** [comment]
@bootstraponline : thanks for responding.

@centic9 : Are you one of the maintainers?

**Comment by centic9** [comment]
My time is very limited to work on this project and only one other dev is active from time to time, so unless someone steps up and agrees to maintain these changes, I am reluctant to merge it likely will require a bit more work and a few iterations until it is "production ready". 

Becoming a committer should be easy, see https://jenkins.io/project/governance/, quote: "Drop us a note in the IRC channel or the dev list, and we’ll set you up with commit access."

**Comment by centic9** [comment]
Nobody showed up after my call for help for more than 2 years so I am closing this for now without merging.

**Pull request closed without merging**